### ISSUE TITLE
Abnormal Reward Curve When Reproducing train_ppo.sh

### ISSUE THREAD

**Issue by Stardust-y** [PeterGriffinJin/Search-R1#97]
Hi, thank you for your great work — it's very inspiring!

I’m trying to reproduce the results using your train_ppo.sh script, but I'm encountering an issue with the reward curve during training.

According to your results, the reward starts to increase around step 80. However, in my case, even after 120+ steps, the reward shows no clear upward trend and fluctuates heavily.

Here are some details of my setup:

I used the default data preprocessing and training configuration.

The only modification I made was reducing all batch size–related hyperparameters by half, since I’m training on an A6000 GPU with limited memory.

Do you have any insights into what might be causing this issue? Is the batch size critical to PPO stability in your setup?

Thanks again for your work, and I look forward to your guidance!

<img width="436" alt="Image" src="https://github.com/user-attachments/assets/aec4e6c6-fbab-4e62-bd4a-423e6b740c1a" />

**Comment by Stardust-y** [comment]
Here're my training setting: 
PYTHONUNBUFFERED=1 python3 -m verl.trainer.main_ppo \
    data.train_files=$DATA_DIR/train.parquet \
    data.val_files=$DATA_DIR/test.parquet \
    data.train_data_num=null \
    data.val_data_num=null \
    data.train_batch_size=256 \
    data.val_batch_size=128 \
    data.max_prompt_length=4096 \
    data.max_response_length=500 \
    data.max_start_length=2048 \
    data.max_obs_length=500 \
    data.shuffle_train_dataloader=True \
    algorithm.adv_estimator=gae \
    actor_rollout_ref.model.path=$BASE_MODEL \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.model.enable_gradient_checkpointing=true \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.actor.optim.lr_warmup_steps_ratio=0.285 \
    actor_rollout_ref.actor.ppo_mini_batch_size=128 \
    actor_rollout_ref.actor.ppo_micro_batch_size=32 \
    actor_rollout_ref.actor.fsdp_config.param_offload=false \
    actor_rollout_ref.actor.fsdp_config.grad_offload=false \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=false \
    +actor_rollout_ref.actor.fsdp_config.model_dtype="bfloat16" \
    actor_rollout_ref.rollout.log_prob_micro_batch_size=64 \
    actor_rollout_ref.rollout.tensor_model_parallel_size=1 \
    actor_rollout_ref.rollout.name=vllm \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.6 \
    actor_rollout_ref.ref.log_prob_micro_batch_size=64 \
    actor_rollout_ref.ref.fsdp_config.param_offload=false \
    +actor_rollout_ref.ref.fsdp_config.model_dtype="bfloat16" \
    actor_rollout_ref.rollout.n_agent=1 \
    actor_rollout_ref.rollout.temperature=1 \
    actor_rollout_ref.actor.state_masking=true \
    critic.optim.lr=1e-5 \
    critic.model.use_remove_padding=True \
    critic.optim.lr_warmup_steps_ratio=0.015 \
    critic.model.path=$BASE_MODEL \
    critic.model.enable_gradient_checkpointing=true \
    critic.ppo_micro_batch_size=8 \
    critic.model.fsdp_config.param_offload=false \
    critic.model.fsdp_config.grad_offload=false \
    critic.model.fsdp_config.optimizer_offload=false \
    +critic.model.fsdp_config.model_dtype="bfloat16" \
    algorithm.kl_ctrl.kl_coef=0.001 \
    algorithm.no_think_rl=false \
    trainer.critic_warmup=0 \
    trainer.logger=['wandb'] \
    +trainer.val_only=false \
    +trainer.val_before_train=false \
    trainer.default_hdfs_dir=null \
    trainer.n_gpus_per_node=4 \
    trainer.nnodes=1 \
    trainer.save_freq=50 \
    trainer.test_freq=100 \
    trainer.project_name=$WAND_PROJECT \
    trainer.experiment_name=$EXPERIMENT_NAME \
    trainer.total_epochs=15 \
    trainer.total_training_steps=1005 \
    trainer.default_hdfs_dir=null \
    trainer.default_local_dir=verl_checkpoints/$EXPERIMENT_NAME \
    max_turns=2 \
    retriever.url="http://127.0.0.1:8000/retrieve" \
    retriever.topk=3 \
    2>&1 | tee $EXPERIMENT_NAME.log

**Comment by seuchenyu** [comment]
i  have  the  same case😂

**Comment by PeterGriffinJin** [comment]
Hi there,

Thank you so much for your interest in our work!

I feel the batch size should not affect the RL so much. Training batch with 256 seems good to me. 

Here are the tips to debug:

- About batch size

"data.train_batch_size" is not the real batch size when doing LLM optimization; you can still keep it as 512, but lower "xxx_micro_batch_size". (See https://verl.readthedocs.io/en/v0.2.x/perf/perf_tuning.html#batch-size-tuning-tips for details)

- fsdp_config.model_dtype

This is another big difference from our default setting. I am not sure if enabling "bfloat16" makes the RL collapse.

Hope that helps, and let me know if you have other questions.

Best,
Bowen

**Comment by Tiagosf00** [comment]
A bit late to the party, but I ran into similar problems reproducing this dataset in another RL framework,  [AReaL](https://github.com/inclusionAI/AReaL), using GRPO.

In my case, the issue turned out to be `actor_rollout_ref.actor.optim.lr_warmup_steps_ratio=0.285`, which was too large when combined with `trainer.total_epochs=15`. The reward stayed flat for many early training steps. After reducing the warmup ratio and training for only 5 epochs, I was able to reproduce the results.
